File: src/types/control.ts
export type ControlType = 'preventive' | 'detective' | 'corrective';
export type ControlStatus = 'planned' | 'implemented' | 'under_review';
export type ControlEffectiveness = 'low' | 'medium' | 'high';

export interface Control {
  id: string;
  title: string;
  description: string;
  controlType: ControlType;
  effectiveness: ControlEffectiveness;
  status: ControlStatus;
  ownerId: string;
  ownerName?: string;
  createdBy: string;
  createdAt: string;
  updatedAt: string;
}

export interface RiskControlMapping {
  id: string;
  riskId: string;
  controlId: string;
  residualLikelihood?: number;
  residualImpact?: number;
  residualRiskScore?: number;
  mappedBy: string;
  mappedAt: string;
}

File: backend/app/schemas/control.py
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, ConfigDict
from pydantic.alias_generators import to_camel
from uuid import UUID
from app.models.control import ControlType, ControlEffectiveness, ControlStatus

class ControlBase(BaseModel):
    title: str
    description: str
    control_type: ControlType
    effectiveness: ControlEffectiveness
    status: ControlStatus = ControlStatus.planned
    linked_risk_id: Optional[UUID] = None
    owner_id: Optional[UUID] = None

    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)

class ControlCreate(ControlBase):
    pass

class ControlUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    control_type: Optional[ControlType] = None
    effectiveness: Optional[ControlEffectiveness] = None
    status: Optional[ControlStatus] = None
    linked_risk_id: Optional[UUID] = None
    owner_id: Optional[UUID] = None

    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)

class ControlInDBBase(ControlBase):
    id: UUID
    created_by: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True, alias_generator=to_camel, populate_by_name=True)

class Control(ControlInDBBase):
    pass
